pl_examples/domain_templates/computer_vision_fine_tuning.py

```diff
@@ -0,0 +1,441 @@
+from torch.utils.data import DataLoader
+from torchvision import models
+from torchvision import transforms
+from torchvision.datasets import ImageFolder
+from torchvision.datasets.utils import download_and_extract_archive
+
+import pytorch_lightning as pl
+from pytorch_lightning import _logger as log
+
+BN_TYPES = (torch.nn.BatchNorm1d, torch.nn.BatchNorm2d, torch.nn.BatchNorm3d)
+DATA_URL = 'https://storage.googleapis.com/mledu-datasets/cats_and_dogs_filtered.zip'
+
+
+#  --- Utility functions ---
+
+
+def _make_trainable(module: torch.nn.Module) -> None:
+    """Unfreeze a given module.
+
+    Args:
+        module (torch.nn.Module): The module to unfreeze
+    """
+    for param in module.parameters():
+        param.requires_grad = True
+    module.train()
+
+
+def _recursive_freeze(module: torch.nn.Module,
+                      train_bn: Optional[bool] = True) -> None:
+    """Freeze the layers of a given module.
+
+    Args:
+        module (torch.nn.Module): The module to freeze
+        train_bn (bool): If True, leave the BatchNorm layers in training mode
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                param.requires_grad = False
+            module.eval()
+        else:
+            # Make the BN layers trainable
+            _make_trainable(module)
+    else:
+        for child in children:
+            _recursive_freeze(module=child, train_bn=train_bn)
+
+
+def freeze(module: torch.nn.Module,
+           n: Optional[int] = None,
+           train_bn: Optional[bool] = True) -> None:
+    """Freeze the layers up to index n (if n is not None).
+
+    Args:
+        module (torch.nn.Module): The module to freeze (at least partially)
+        n (int): Max depth at which we stop freezing the layers. By default,
+        train_bn (bool): If True, leave the BatchNorm layers in training mode
+    """
+    children = list(module.children())
+    n_max = len(children) if n is None else int(n)
+
+    for child in children[:n_max]:
+        _recursive_freeze(module=child, train_bn=train_bn)
+
+    for child in children[n_max:]:
+        _make_trainable(module=child)
+
+
+def filter_params(module: torch.nn.Module,
+                  train_bn: Optional[bool] = True) -> Generator:
+    """Yield the trainable parameters of a given module.
+
+    Args:
+        module (torch.nn.Module): A given module
+        train_bn (bool): If True, leave the BatchNorm layers in training mode
+
+    Returns:
+        Generator
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                if param.requires_grad:
+                    yield param
+    else:
+        for child in children:
+            filter_params(module=child, train_bn=train_bn)
```

Comment:
I think this line should actually be: 
`yield from filter_params(module=child, train_bn=train_bn)`

This way the yielded parameters from a nested call actually get back to the _optimizer.add_param_group_ otherwise (as it is now) they just get "lost" before being returned to the initial call.

Reply:
Thank you for pointing this out ! I missed it. It's now fixed.